ddr/lib/ddr-ir/Macro.cpp

```diff
@@ -21,53 +21,962 @@
+			} else if ((fixedWidthTypes + 1) == found) {
+				*bitWidth = 16;
+				*isSigned = true;
+				validTypes = true;
+			} else if ((fixedWidthTypes + 2) == found) {
+				*bitWidth = 32;
+				*isSigned = true;
+				validTypes = true;
+			} else if ((fixedWidthTypes + 3) == found) {
+				*bitWidth = 64;
+				*isSigned = true;
+				validTypes = true;
+			} else if ((fixedWidthTypes + 4) == found) {
+				*bitWidth = 8;
+				*isSigned = false;
+				validTypes = true;
+			} else if ((fixedWidthTypes + 5) == found) {
+				*bitWidth = 16;
+				*isSigned = false;
+				validTypes = true;
+			} else if ((fixedWidthTypes + 6) == found) {
+				*bitWidth = 32;
+				*isSigned = false;
+				validTypes = true;
+			} else if ((fixedWidthTypes + 7) == found) {
+				*bitWidth = 64;
+				*isSigned = false;
+				validTypes = true;
+			}
+			if (validTypes) {
+				return validTypes;
+			}
+		}
+
+		/* C standard allows types and type modifiers in any order,
+		 * so we count the number of each type and verify it is correct
+		 */
+		size_t numChar = std::count(types.begin(), types.end(), "char");
+		size_t numShort = std::count(types.begin(), types.end(), "short");
+		size_t numInt = std::count(types.begin(), types.end(), "int");
+		size_t numLong = std::count(types.begin(), types.end(), "long");
+		size_t numSigned = std::count(types.begin(), types.end(), "signed");
+		size_t numUnsigned = std::count(types.begin(), types.end(), "unsigned");
+
+		/* total count of valid types should equal the size of types vector */
+		if (types.size() == (numChar + numShort + numInt + numLong + numSigned + numUnsigned)) {
+			if (numUnsigned > 0) {
+				*isSigned = false;
+			} else {
+				*isSigned = true;
+			}
+			/* (1 >= numSigned + numUnsigned) checks that there can only be one
+			 *  of 'unsigned' or 'signed' in the types
+			 */
+			if ((1 == numChar) && (1 >= numSigned + numUnsigned) && (0 == numShort)
+					&& (0 == numInt) && (0 == numLong)) {
+				*bitWidth = sizeof(char) * 8;
+				validTypes = true;
+			} else if ((1 == numShort) && (1 >= numSigned + numUnsigned)
+					&& (0 == numChar) && (1 >= numInt) && (0 == numLong)) {
+				*bitWidth = sizeof(short) * 8;
+				validTypes = true;
+			} else if ((1 == numInt) && (1 >= numSigned + numUnsigned)
+					&& (0 == numChar) && (0 == numShort) && (0 == numLong)) {
+				*bitWidth = sizeof(int) * 8;
+				validTypes = true;
+			} else if ((1 == numLong) && (1 >= numSigned + numUnsigned)
+					&& (0 == numChar) && (0 == numShort) && (1 >= numInt)) {
+				*bitWidth = sizeof(long) * 8;
+				validTypes = true;
+			} else if ((2 == numLong) && (1 >= numSigned + numUnsigned)
+					&& (0 == numChar) && (0 == numShort) && (1 >= numInt)) {
+				*bitWidth = sizeof(long long) * 8;
+				validTypes = true;
+			} else if ((1 >= numSigned + numUnsigned) && (0 == numChar)
+					&& (0 == numShort) && (0 == numInt) && (0 == numLong)) {
+				*bitWidth = 0;
+				validTypes = true;
+			}
 		}
+	}
+	if (!validTypes) {
+		_cursor = typecastStart - 1;
```

Comment:
Why the `- 1`? Shouldn't it restore _cursor to the value on entry?

Reply:
It restores `_cursor` to the `(` symbol before the string of types.

Reply:
That would be reasonable if the scanner (rather than the parser) considered the opening parenthesis.